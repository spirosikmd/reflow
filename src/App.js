import React, { PureComponent } from 'react';
import Flow from './Flow';
import updates from './updates';
import { Button, Grid } from 'semantic-ui-react';

class App extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      updates: [],
      step: 0,
    };

    this.handleForwardClick = this.handleForwardClick.bind(this);
    this.handleBackwardClick = this.handleBackwardClick.bind(this);
  }

  isTheEnd(state, updates) {
    return state.step === updates.length;
  }

  isTheStart(state) {
    return state.step === 0;
  }

  handleForwardClick(state, updates) {
    if (this.isTheEnd(state, updates)) {
      return;
    }
    const newUpdates = [...state.updates, updates[state.step]];
    const step = state.step + 1;
    this.setState({ updates: newUpdates, step });
  }

  handleBackwardClick(state) {
    if (this.isTheStart(state)) {
      return;
    }
    const newUpdates = state.updates.slice(0, state.updates.length - 1);
    const step = state.step - 1;
    this.setState({ updates: newUpdates, step });
  }

  render() {
    return (
      <div>
        <Grid>
          <Grid.Row>
            <Grid.Column width={4}>
              <Button
                icon="backward"
                onClick={() => this.handleBackwardClick(this.state)}
              />
              <Button
                icon="forward"
                onClick={() => this.handleForwardClick(this.state, updates)}
              />
            </Grid.Column>
            <Grid.Column width={12} textAlign="center">
              <Flow updates={this.state.updates} />
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </div>
    );
  }
}

export default App;
