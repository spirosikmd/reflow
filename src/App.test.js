import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

it('App renders', () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('App isTheEnd returns true when step equals updates length', () => {
  const app = new App();
  expect(app.isTheEnd({ step: 2 }, [{}, {}])).toBe(true);
});

it('App isTheEnd returns false when step does not equal updates length', () => {
  const app = new App();
  expect(app.isTheEnd({ step: 1 }, [{}, {}])).toBe(false);
});

it('App isTheStart returns true when step equals 0', () => {
  const app = new App();
  expect(app.isTheStart({ step: 0 })).toBe(true);
});

it('App isTheStart returns true when step does not equal 0', () => {
  const app = new App();
  expect(app.isTheStart({ step: 2 })).toBe(false);
});

it('App handleForwardClick does not call setState when is the end', () => {
  const app = new App();
  app.setState = jest.fn();
  app.isTheEnd = jest.fn().mockReturnValue(true);
  app.handleForwardClick();
  expect(app.setState).not.toHaveBeenCalled();
});

it('App handleForwardClick calls setState with new updates and step', () => {
  const app = new App();
  app.setState = jest.fn();
  app.isTheEnd = jest.fn().mockReturnValue(false);
  app.handleForwardClick({ step: 1, updates: [{}] }, [{}, {}]);
  expect(app.setState).toHaveBeenCalledWith({
    updates: [{}, {}],
    step: 2,
  });
});

it('App handleBackwardClick does not call setState when is the start', () => {
  const app = new App();
  app.setState = jest.fn();
  app.isTheStart = jest.fn().mockReturnValue(true);
  app.handleBackwardClick();
  expect(app.setState).not.toHaveBeenCalled();
});

it('App handleBackwardClick calls setState with new updates and step', () => {
  const app = new App();
  app.setState = jest.fn();
  app.isTheStart = jest.fn().mockReturnValue(false);
  app.handleBackwardClick({ step: 2, updates: [{}, {}] });
  expect(app.setState).toHaveBeenCalledWith({
    updates: [{}],
    step: 1,
  });
});

it('App handlerResetClick resets updates and step', () => {
  const app = new App();
  app.setState = jest.fn();
  app.handlerResetClick();
  expect(app.setState).toHaveBeenCalledWith({ step: 0, updates: [] });
});

it('App getUpdateIcon returns correct icon given on update type', () => {
  const app = new App();
  expect(app.getUpdateIcon({ type: 'hotfix' })).toBe('bug');
  expect(app.getUpdateIcon({ type: 'master' })).toBe('rocket');
  expect(app.getUpdateIcon({ type: 'develop' })).toBe('code');
});
