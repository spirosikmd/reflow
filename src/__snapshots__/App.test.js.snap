exports[`test App renders 1`] = `
<div>
  <div
    className="ui grid">
    <div
      className="row">
      <div
        className="five wide column">
        <button
          className="ui icon button"
          onClick={[Function]}
          tabIndex={undefined}>
          <i
            aria-hidden="true"
            className="backward icon" />
           
        </button>
        <button
          className="ui icon button"
          onClick={[Function]}
          tabIndex={undefined}>
          <i
            aria-hidden="true"
            className="forward icon" />
           
        </button>
        <button
          className="ui button"
          onClick={[Function]}
          tabIndex={undefined}>
          reset
        </button>
        <div
          className="ui list"
          role="list" />
      </div>
      <div
        className="center aligned eleven wide column">
        <div>
          <svg
            height="700"
            viewBox="0 0 400 700"
            width="100%">
            <text
              x={60}
              y="20">
              feature
            </text>
            <text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              x={145}
              y="20">
              develop
            </text>
            <text
              x={210}
              y="20">
              release
            </text>
            <text
              x={265}
              y="20">
              hotfixes
            </text>
            <text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              x={330}
              y="20">
              master
            </text>
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={50}
              x2={50}
              y1={30}
              y2={680} />
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={110}
              x2={110}
              y1={30}
              y2={680} />
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={170}
              x2={170}
              y1={30}
              y2={680} />
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={230}
              x2={230}
              y1={30}
              y2={680} />
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={290}
              x2={290}
              y1={30}
              y2={680} />
            <line
              style={
                Object {
                  "stroke": "rgb(208, 208, 208)",
                  "strokeWidth": 1,
                }
              }
              x1={350}
              x2={350}
              y1={30}
              y2={680} />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
